packages/react/react/src/use-resource.ts

```diff
@@ -1,140 +1,139 @@
+import { callerStack } from "@starbeam/debug";
+import { LIFETIME, TIMELINE } from "@starbeam/timeline";
 import {
+  type IntoResource,
+  type Reactive,
   type ResourceBlueprint,
-  Cell,
-  PolledFormulaFn,
-  Resource,
-  TIMELINE,
-} from "@starbeam/core";
-import type { Description } from "@starbeam/debug";
-import { descriptionFrom } from "@starbeam/debug";
-import { LIFETIME } from "@starbeam/timeline";
+  Factory,
+  Formula,
+  Marker,
+} from "@starbeam/universal";
 import {
   unsafeTrackedElsewhere,
   useLifecycle,
 } from "@starbeam/use-strict-lifecycle";
-import { useState } from "react";
-
-function createReactiveResource<T>(
-  resource: () => ResourceBlueprint<T>,
-  deps: unknown[] | undefined,
-  description: string | Description | undefined
-): Resource<T> {
-  const desc = descriptionFrom({
-    type: "resource",
-    api: "useReactiveSetup",
-    fromUser: description,
-  });
-
-  const [, setNotify] = useState({});
-
-  const instance = useLifecycle({ resource, deps }, (lifecycle) => {
-    let lastDeps = deps;
-    let setup = false;
-
-    const owner = {};
-
-    let currentResource = resource().create({
-      owner,
-    });
 
-    const resourceCell = Cell(currentResource);
+import { useNotify } from "./use-reactive.js";
+import { useComponent } from "./use-setup.js";
+
+export type UseFactory<T, D extends undefined> =
+  | ResourceBlueprint<T, D>
+  | (() => IntoResource<T>);
+
+export function use<T>(
+  factory: ResourceBlueprint<T> | (() => ResourceBlueprint<T>),
+  dependencies?: unknown[]
+): T;
+// export function use<T>(
+//   factory: ResourceBlueprint<T, UNINITIALIZED>
+// ): T | undefined;
+// export function use<T>(
+//   factory: UseFactory<T, UNINITIALIZED>,
+//   dependencies: unknown[]
+// ): T | undefined;
+// export function use<T>(
+//   factory: UseFactory<T, never>,
+//   dependencies: unknown[]
+// ): T;
+// export function use<T>(
+//   factory: UseFactory<T, UNINITIALIZED> | UseFactory<T, never>,
+//   options: { initial: T },
+//   dependencies: unknown[]
+// ): T;
+
+export function use<T>(
+  factory: UseFactory<T, undefined>,
+  options?: { initial: T } | unknown[],
+  dependencies?: unknown[]
+): T | undefined {
+  const value = createResource(factory, options, dependencies);
+
+  return unsafeTrackedElsewhere(() => value.current);
```

Comment:
This doesn't affect the semantics of the system. What it does is allow a consumption to happen in a React render context that's not inside of an auto-tracking frame. Normally, we consider that an error (and tell you to wrap your code in `useReactive` or `useResource`), to ensure that reactive consumptions properly update the output.

The core primitives (`useReactive` and `useResource`) are responsible for subscribing to the reactive value in question and notifying react when an update occurs, and then return the value of the reactive.

In other words, the core primitives themselves need to get the current value of a reactive outside of a tracking context **because they, themselves are already taking care of the subscription and notification**.

I want to reiterate that this is _doesn't affect reactive semantics_, and exists entirely to provide proactive errors when a React component consumes reactive values outside of a tracking context.

One additional note: The Preact renderer automatically wraps all components in a tracking frame (using Preact's options API), so this concern doesn't come up there. Since there's no way to consume reactive values outside of a tracking frame when using `@starbeam/preact`, everything just works.

Reply:
ah interesting, so unsafeTrackedElsewhere exists solely because of React reasons?
I do like the focus on proactive errors!

would any other frameworks need it? (that you know of?) :upside_down_face:

Reply:
React is the only framework (so far) that has needed us to ask users to manually insert a tracking frame around reactive consumption during rendering. But it's definitely possible that another framework will need the same thing.